File: JavaFundamentals/ExamProblems/ExamPractice/src/org/softuni/PopulationCounter.java
package org.softuni;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Scanner;

// Advanced C# Exam 19 July 2015

public class PopulationCounter {

    public static void main(String[] args) {

        Scanner in = new Scanner(System.in);
        Map<String, Long> countries = new LinkedHashMap<>();
        LinkedHashMap<String, LinkedHashMap<String, Long>> cities = new LinkedHashMap<>();
        String input = in.nextLine();
        while (!input.equals("report")) {
            String[] parameters = input.split("[|]");
            String city = parameters[0];
            String country = parameters[1];
            long population = Long.parseLong(parameters[2]);

            if (!countries.containsKey(country)) {
                countries.put(country, 0L);
                cities.put(country, new LinkedHashMap<>());
            }
            countries.put(country, countries.get(country) + population);
            if (!cities.get(country).containsKey(city)) {
                cities.get(country).put(city, 0L);
            }
            cities.get(country).put(city, cities.get(country).get(city) + population);

            input = in.nextLine();
        }

        countries.entrySet().stream()
                .sorted((e1, e2) -> e2.getValue().compareTo(e1.getValue()))
                .forEachOrdered(country -> {
                    System.out.printf("%s (total population: %d)\r\n", country.getKey(), country.getValue());
                    cities.get(country.getKey()).entrySet().stream()
                            .sorted((e1, e2) -> e2.getValue().compareTo(e1.getValue()))
                            .forEachOrdered(city -> System.out.printf("=>%s: %d\r\n", city.getKey(), city.getValue()));
                });
    }
}

File: C#-WebDeveloper-3.0/Advanced-C#-May-2016/Exercises/SetsAndDictionaries/PopulationCounter/PopulationCounter.cs
namespace PopulationCounter
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    class PopulationCounter
    {
        static void Main(string[] args)
        {
            var countries = new Dictionary<string, long>();
            var cities = new Dictionary<string, Dictionary<string, long>>();
            string input = Console.ReadLine();
            while (input != "report")
            {
                string[] parameters = input.Split(new[] { '|' }, StringSplitOptions.RemoveEmptyEntries);
                string city = parameters[0];
                string country = parameters[1];
                long population = long.Parse(parameters[2]);

                if (!countries.ContainsKey(country))
                {
                    countries[country] = 0L;
                    cities[country] = new Dictionary<string, long>();
                }

                countries[country] += population;

                if (!cities[country].ContainsKey(city))
                {
                    cities[country][city] = 0L;
                }

                cities[country][city] += population;

                input = Console.ReadLine();
            }

            foreach (var country in countries.OrderByDescending(pair => pair.Value))
            {
                Console.WriteLine($"{country.Key} (total population: {country.Value})");

                foreach (var city in cities[country.Key].OrderByDescending(pair => pair.Value))
                {
                    Console.WriteLine($"=>{city.Key}: {city.Value}");
                }
            }
        }
    }
}
